openTSNE/nearest_neighbors.py

```diff
@@ -445,3 +445,89 @@ def query(self, query, k):
 
         return indices, distances
 
+
+class HNSW(KNNIndex):
+    VALID_METRICS = [
+        "cosine",
+        "euclidean",
+        "dot",
+        "l2",
+        "ip",
+    ]
+
+    def __init__(self, *args, **kwargs):
+        try:
+            from hnswlib import Index  # pylint: disable=unused-import,unused-variable
+        except ImportError:
+            raise ImportError(
+                "Please install hnswlib: `conda install -c conda-forge "
+                "hnswlib` or `pip install hnswlib`."
+            )
+        super().__init__(*args, **kwargs)
+
+    def build(self, data, k: int):
+        timer = utils.Timer(
+            f"Finding {k} nearest neighbors using HNSWlib approximate search using "
+            f"{self.metric} distance...",
+            verbose=self.verbose,
+        )
+        timer.__enter__()
+
+        from hnswlib import Index
+
+        hnsw_space = {
+            "cosine": "cosine",
+            "dot": "ip",
+            "euclidean": "l2",
+            "ip": "ip",
+            "l2": "l2"
+        }[self.metric]
+
+        self.index = Index(space=hnsw_space, dim=data.shape[1])
+
+        metric_params = {
+            'max_elements': data.shape[0],
+            'ef_construction': 200,
+            'M': 16,
+            'random_seed': self.random_state or 100
+        }
+        if self.metric_params is not None:
+            metric_params.update(self.metric_params)
+
+        # Initialize HNSW Index
+        self.index.init_index(**metric_params)
+
+        # Build index tree from data
+        self.index.add_items(data, num_threads=self.n_jobs)
+
+        # Set ef parameter for (ideal) precision/recall
+        self.index.set_ef(2 * k)
+
+        # Query for kNN
+        indices, distances = self.index.knn_query(data, k=k + 1, num_threads=self.n_jobs)
+
+        # Stop timer
+        timer.__exit__()
+
+        # return indices and distances, skip first entry, which is always the point itself
+        return indices[:, 1:], distances[:, 1:]
+
+    def query(self, query, k):
+        timer = utils.Timer(
+            f"Finding {k} nearest neighbors in existing embedding using HNSWlib "
+            f"approximate search...",
+            self.verbose,
+        )
+        timer.__enter__()
+
+        # Set ef parameter for (ideal) precision/recall
+        self.index.set_ef(2 * k)
```

Comment:
This is a fixed setting I typically use. `hnswlib` recommends any value within `k < ef < index size`.  
https://github.com/nmslib/hnswlib/blob/master/ALGO_PARAMS.md

Reply:
Hmm, I don't know, I think we definitely need this check. Using high perplexity values is useful in a lot of cases, so we need a lot of neighbors. So we definitely can't have this triggering an error. I would allow any k, actually. Why would we limit the number to (2*k). In that case it wouldn't make sense to use HNSW at all, since we'd need the entire distance matrix. But I'd leave that up to the user. Maybe a warning...

Reply:
I checked it, `hnswlib` doesn't throw any errors. I limited the highest number that can be set anyway just to be sure nothing weird happens under the hood.